### PR TITLE
Preconditioned CG_LANCZOS

### PR DESCRIPTION
**Pull request by amontoison**
#52 
I preconditioned those methods the same way that I did in `DQGMRES` in `DIOM`.  Lanczos orthogonalization is just a special case of Arnoldi orthogonalization, so we can add a preconditioner the same way.

However I don't really understood how we have the relations for updating `p` and `x`. I only know how to derived relations with a LU factorization of `Tₘ` and  it's quite different.  I suppose that we modified them with the help of the relation `σⱼ² (δⱼ - ωⱼ₋₁ / γⱼ₋₁) = pⱼᵀ A pⱼ`.

### CODE REVIEW

File: src/cg_lanczos.jl [hunk 1]
```diff
@@ -51,29 +54,30 @@ function cg_lanczos(A :: AbstractLinearOperator, b :: AbstractVector{T};
   verbose && @printf("%5d  %8.1e\n", iter, rNorm);
 
   indefinite = false;
-  solved = rNorm <= ε;
-  tired = iter >= itmax;
+  solved = rNorm ≤ ε;
+  tired = iter ≥ itmax;
```
Comment (dpo) [inline src/cg_lanczos.jl]: Same comment here as in #75. Please keep a pull request to a single topic. It's great to submit another pull requests with syntactic updates. Otherwise, it takes me much longer to review a pull request like this one. Thank you.

Comment (Xx-fractale-xX) [reply src/cg_lanczos.jl]: I cleaned each pull request and created new pull requests when it was about another topic. It will be easier to review now.

File: src/cg_lanczos.jl [hunk 2]
```diff
@@ -18,10 +18,13 @@ symmetric linear system
   Ax = b
 
 The method does _not_ abort if A is not definite.
+
+This implementation allows a right preconditioner M.
```
Comment (dpo) [inline src/cg_lanczos.jl]: A left preconditioner would be more interesting here. If *M* is symmetric and positive definite, we really want to do centered preconditioning, i.e., *M<sup>1/2</sup> A M<sup>1/2</sup> x = M<sup>1/2</sup> b*. When you write out CG_Lanczos for this system, you can transform it so only *M* is needed (not *M<sup>1/2</sup>*) and everything happens as if you were solving *MAx = Mb*.

If you look at standard CG with preconditioner, you'll see that the first thing we do is precondition the initial residual (i.e., *b*): https://github.com/JuliaSmoothOptimizers/Krylov.jl/blob/master/src/cg.jl#L33. CG_Lanczos and CG should be perfectly equivalent (on paper).

Comment (Xx-fractale-xX) [reply src/cg_lanczos.jl]: I updated it, it's a right preconditioner now for both methods.

Comment (dpo) [reply src/cg_lanczos.jl]: Right or left?

Comment (Xx-fractale-xX) [reply src/cg_lanczos.jl]: Left, I made a mistake in my last message.

File: src/cg_lanczos.jl [hunk 3]
```diff
@@ -18,22 +18,27 @@ symmetric linear system
...
 
   n = size(b, 1);
   (size(A, 1) == n & size(A, 2) == n) || error("Inconsistent problem size");
   verbose && @printf("CG Lanczos: system of %d equations in %d variables\n", n, n);
 
   # Initial state.
   x = zeros(T, n);
-  β = @knrm2(n, b)
+  r₀ = M * b
+  β = @knrm2(n, r₀)
```
Comment (dpo) [inline src/cg_lanczos.jl]: I think this should be `@kdot(n, r₀, b)`.

File: src/cg_lanczos.jl [hunk 4]
```diff
@@ -119,13 +128,14 @@ function cg_lanczos_shift_seq(A :: AbstractLinearOperator, b :: AbstractVector{T
   # Initial state.
   ## Distribute x similarly to shifts.
   x = zeros(Tb, n, nshifts);
-  β = @knrm2(n, b)
+  r₀ = M * b
+  β = @knrm2(n, r₀)
```
Comment (dpo) [inline src/cg_lanczos.jl]: Same here.

File: src/cg_lanczos.jl [hunk 5]
```diff
@@ -161,21 +171,29 @@ function cg_lanczos_shift_seq(A :: AbstractLinearOperator, b :: AbstractVector{T
   # Main loop.
   while ! (solved || tired)
     # Form next Lanczos vector.
-    v_next = A * v;
+    Mv = M * v
+    if Mv === v
+      ρ = 1
+    else
+      ρ = @kdot(n, v, Mv)
```
Comment (dpo) [inline src/cg_lanczos.jl]: Won't `@kdot` return 1 when `Mv === v`?

File: src/cg_lanczos.jl [hunk 6]
```diff
@@ -161,21 +171,29 @@ function cg_lanczos_shift_seq(A :: AbstractLinearOperator, b :: AbstractVector{T
...
+      @. v_prev = v
     end
     β = @knrm2(n, v_next)
-    v = v_next / β;
+    @. v = v_next / β;
 
-    # Check curvature: v'(A + sᵢI)v = v'Av + sᵢ ‖v‖² = δ + sᵢ because ‖v‖ = 1.
-    # It is possible to show that σⱼ² (δⱼ + sᵢ - ωⱼ₋₁ / γⱼ₋₁) = pⱼᵀ (A + sᵢ I) pⱼ.
+    # Check curvature: vᵗ M(A + sᵢI) v = vᵗMAv + sᵢ ‖M½v‖² = δ + ρ * sᵢ.
+    # With M = I, ρ = ‖v‖² = 1.
```
Comment (dpo) [inline src/cg_lanczos.jl]: More generally, *v'Mv = ‖v‖<sub>M</sub>² = ‖M<sup>½</sup>v‖² = 1*.

File: src/cg_lanczos.jl (hunk 3 above)
Comment (Xx-fractale-xX) [reply src/cg_lanczos.jl]: We want that `βMv₁ = b with ‖v₁‖_M = 1`
`βv₁ = M⁻¹b` and by product `β²<v₁,Mv₁> = bᵗM⁻¹b`.
So `β = sqrt(bᵗM⁻¹b)`.

Ok, I see the problem, `β = sqrt(@kdot(n, r₀, b))`

Comment (dpo) [reply src/cg_lanczos.jl]: You're right.

**Comment by dpo** [comment]
This needs to be rebased.

**Comment by amontoison** [comment]
With the new `opEye`, I will update this pull request next week.

**Comment by amontoison** [comment]
It's seem that the expression of `‖rₖ‖₂` is different in `CG_LANCZOS_SHIFT_SEQ` with a preconditioner. I will add it later in an other pull request.

**Comment by amontoison** [comment]
I was reading the original article `Fast CG-Based Methods for Tikhonov-Phillips Regularization` and I found some mistakes in my indices.
it should be correctly preconditioned and documented now.

**Comment by dpo** [comment]
Thank you!
